### ISSUE TITLE
Finishing Code

### ISSUE THREAD

**Issue by PeterPilley** [ygrek/mldonkey#15]
ygrek I note in your code it looks like you have not finished all parts of this api, is it possible to finish?

I have an ongoing project that I am trying to base on mldonkey and your json additions could be really useful but I would like to expose all functions. Are you still developing this?

**Comment by ygrek** [comment]
It is definitely possible. I need feedback on the existing api and what parts do you need most at this point. Also be ready that some things might change, no compatibility guarantee for now.

**Comment by PeterPilley** [comment]
OKay I will start testing and respond back. 

So from the API I can see that the search function, download functions have been exposed unless I missed any other functions.

Can you expose the settings functions?

Are all functions exposed in the API file or are the in the driverInteractive file?

**Comment by ygrek** [comment]
Noted.
Everything is in `src/daemon/driver/driverApi.ml`
Also there is `docs/api.md`

**Comment by PeterPilley** [comment]
The api listens on a seperate interface port correct? 

Where is this established

**Comment by ygrek** [comment]
It listens on the same port as webui - see https://github.com/ygrek/mldonkey/blob/api/docs/api.md
Why would anyone want to download all search results? Anyway via api just going through all results and issuing separate request for each should work

**Comment by PeterPilley** [comment]
Yeah I was thinking the same thing just iterate the results. okay I saw it in the .md file but I missed it in the code it makes sense now. 

basically what I am would like to achieve in my project is present the same functions as mldonkey telnet interface presents http://mldonkey.sourceforge.net/MLdonkeyCommandsExplained.

**Comment by PeterPilley** [comment]
I have isntalled the api branch to test however I cannot access the interface. I am instead presented with an exist web interface.

if I try and access localhost:4080/api/1/ it fails

I configured as per instructions and can see driverAPI has compiled.

**Comment by PeterPilley** [comment]
Okay have tested the api over the weekend very nice, I am adding the downloaders function and it is called from commFile.file_downloaders I am geussing I am needing to present it as list.map I am not a caml programmer can you give some advise

**Comment by PeterPilley** [comment]
so for me the functions I would like to expose from the original web interface would be the 

- vc  or view clients option this looks to be called from commonClient as clients_get_all() and seems to expose both the uploaders and the downloaders functions. By default it is vc all from the telnet interface and that is treated as an argument. 
it would be awesome to be able to expose the entire function and be able to expose the vc [num] where you can see an individual source data 

is this possible?

**Comment by ygrek** [comment]
see example at https://github.com/ygrek/mldonkey/commit/8dceb41fe71c1c040208d40d44a2104db694e125

**Comment by PeterPilley** [comment]
Testing now.

**Comment by PeterPilley** [comment]
Have country code and IP exposed now, thanks for example.

Could the downloads call be modified to specify a specific user to download from?

**Comment by ygrek** [comment]
> Have country code and IP exposed now, thanks for example.

Do you mean you have added it? Please share the patch/branch/PR.

> Could the downloads call be modified to specify a specific user to download from?

I don't quite understand the question, what do you have in mind?

**Comment by PeterPilley** [comment]
Okay I am trying to work on your download call to add the hash of the file that is returned for the file that is downloading

I can see this is the type in driverApiType.ml

let of_file f =
  let f = CommonFile.file_info f in
  {
    T.file_id = f.file_num;
    file_name = f.file_name;
    file_size = f.file_size;
    file_downloaded = f.file_downloaded;
    file_download_rate = f.file_download_rate;
    file_network = net_shortname f;
    file_state = f.file_state;
  }

I have tried adding a file_print_ed2k_link and others with no success. 

this is the same for the client id for the clients call 

any ideas?

Also do you have an email to enable easier discussion

**Comment by ygrek** [comment]
See email in commits. Also there is #mldonkey irc on freenode.

**Comment by PeterPilley** [comment]
Okay sure I will go through the commits.

**Comment by PeterPilley** [comment]
Have sent email did you get it?

I am looking at the new clients call that you did can you advise on how to include the results from donkeyinteractive or any of the interactives.

Specifically I need to associate the client with the downloading file I thought it was the file_queue that you have as queue_length but I cant see it there, and I want to expost the client MD4 (will try and do the same for the gnutellainteractive as well).

Also is it possible for the search results to show who the result is coming from like in gnutella you have the locations stored within remotefiledesc as well as the alt locations. can results.json contain an array of the result locations?
